Encryption.md

```diff
@@ -19,216 +19,544 @@
+is a union - the actual structure is chosen depending on whether the column is encrypted with the 
+footer key, or with a column-specific key. For the latter, a key metadata can be specified.
+
+```c
+struct EncryptionWithFooterKey {
+}
+
+struct EncryptionWithColumnKey {
+  /** Column path in schema **/
+  1: required list<string> path_in_schema
+  
+  /** Retrieval metadata of the column-specific key **/
+  2: optional binary column_key_metadata
+}
+
+union ColumnCryptoMetaData {
+  1: EncryptionWithFooterKey ENCRYPTION_WITH_FOOTER_KEY
+  2: EncryptionWithColumnKey ENCRYPTION_WITH_COLUMN_KEY
+}
+
+struct ColumnChunk {
+...
+  /** Crypto metadata of encrypted columns **/
+  8: optional ColumnCryptoMetaData crypto_meta_data
+}
+```
+
+The row group ordinal, required for AAD_Suffix calculation, is set in the RowGroup structure:
+
+```c
+struct RowGroup {
+...
+  /** Row group ordinal in the file **/
+  7: optional i16 ordinal
+}
+```
+
+### 5.1 Protection of column metadata
+The Parquet file footer, and its nested structures, contain sensitive information - ranging 
 from a secret data (column statistics) to other information that can be exploited by an 
-attacker (e.g. schema, num_values, key_value_metadata, column data offset and size, encoding and crypto_meta_data). 
-This information is automatically protected when the footer and secret columns are encrypted 
-with the same key. In other cases - when column(s) and the footer are encrypted with 
-different keys; or column(s) are encrypted and the footer is not - an extra measure is 
-required to protect the column-specific information in the file footer. In these cases, 
-the column-specific information (kept in `ColumnMetaData` structures) is split from the 
-footer, by utilizing the `required i64 file_offset` parameter in the `ColumnChunk` 
-structure. This allows to serialize each `ColumnMetaData` structure separately, and encrypt 
-it with a column-specific key, thus protecting the column stats and other metadata. 
-
-### Encrypted footer mode
-
+attacker (e.g. schema, num_values, key_value_metadata, encoding 
+and crypto_meta_data). This information is automatically protected when the footer and 
+secret columns are encrypted with the same key. In other cases - when column(s) and the 
+footer are encrypted with different keys; or column(s) are encrypted and the footer is not, 
+an extra measure is required to protect the column-specific information in the file footer. 
+In these cases, the `ColumnMetaData` structures are Thrift-serialized separately and encrypted 
+ with the AES GCM cipher and a column-specific key, thus protecting the column stats and 
+ other metadata. The encryption buffer is kept in an `optional binary encrypted_column_metadata` 
+ field in the `ColumnChunk`.
+
+```c
+struct ColumnChunk {
+...
+  
+  /** Column metadata for this chunk.. **/
+  3: optional ColumnMetaData meta_data
```

Comment:
Do we know how older readers act when `meta_data` is null?

Reply:
Nevermind, I see that `meta_data` is still required in plaintext footer mode. Glad to see you already caught it!